PWGCF/GenericFramework/Tasks/flowGenericFramework.cxx

```diff
@@ -390,58 +381,64 @@ struct GenericFramework {
     } else if constexpr (framework::has_type_v<aod::mcparticle::McCollisionId, typename TrackObject::all_columns>) {
-      pt = track.pt();
-      eta = track.eta();
-      phi = track.phi();
-      if (!track.isPhysicalPrimary() || eta < etalow || eta > etaup || pt < ptlow || pt > ptup)
+      if (!track.isPhysicalPrimary() || track.eta() < etalow || track.eta() > etaup || track.pt() < ptlow || track.pt() > ptup)
         return;
-      if (cfgFillQA) {
-        QAregistry.fill(HIST("phi_gen"), phi);
-        QAregistry.fill(HIST("eta_gen"), eta);
-        QAregistry.fill(HIST("pt_gen"), pt);
-      }
+      if (cfgFillQA)
+        FillQA<kGen>(track.phi(), track.eta(), track.pt());
+      FillGFW(track.phi(), track.eta(), track.pt(), 1.);
     } else {
-      pt = track.pt();
-      eta = track.eta();
-      phi = track.phi();
       if (cfgFillWeights)
-        fWeights->Fill(phi, eta, vtxz, pt, centrality, 0);
-      if (!setCurrentParticleWeights(weff, wacc, phi, eta, pt, vtxz))
+        fWeights->Fill(track.phi(), track.eta(), vtxz, track.pt(), centrality, 0);
+      if (!setCurrentParticleWeights(weff, wacc, track.phi(), track.eta(), track.pt(), vtxz))
         return;
-      registry.fill(HIST("phi_eta_vtxZ_corrected"), phi, eta, vtxz, wacc);
-      if (cfgFillQA) {
-        QAregistry.fill(HIST("phi"), phi);
-        QAregistry.fill(HIST("eta"), eta);
-        QAregistry.fill(HIST("pt_dcaXY_dcaZ"), pt, track.dcaXY(), track.dcaZ());
-      }
+      registry.fill(HIST("phi_eta_vtxZ_corrected"), track.phi(), track.eta(), vtxz, wacc);
+      if (cfgFillQA)
+        FillQA<kData>(track.phi(), track.eta(), track.pt(), track.dcaXY(), track.dcaZ());
+      FillGFW(track.phi(), track.eta(), track.pt(), weff * wacc);
     }
+  }
+
+  template <typename T>
+  inline void FillGFW(T phi, T eta, T pt, float w)
+  {
     bool WithinPtPOI = (ptpoilow < pt) && (pt < ptpoiup); // within POI pT range
     bool WithinPtRef = (ptreflow < pt) && (pt < ptrefup); // within RF pT range
     if (WithinPtRef)
-      fGFW->Fill(track.eta(), fPtAxis->FindBin(pt) - 1, phi, wacc * weff, 1);
+      fGFW->Fill(eta, fPtAxis->FindBin(pt) - 1, phi, w, 1);
     if (WithinPtPOI)
-      fGFW->Fill(track.eta(), fPtAxis->FindBin(pt) - 1, phi, wacc * weff, 2);
+      fGFW->Fill(eta, fPtAxis->FindBin(pt) - 1, phi, w, 2);
     if (WithinPtPOI && WithinPtRef)
-      fGFW->Fill(track.eta(), fPtAxis->FindBin(pt) - 1, phi, wacc * weff, 4);
+      fGFW->Fill(eta, fPtAxis->FindBin(pt) - 1, phi, w, 4);
+    return;
+  }
+
+  template <datatype dt, typename T>
+  inline void FillQA(T phi, T eta, T pt, T dcaxy = 0, T dcaz = 0)
+  {
+    if (dt == kGen) {
+      QAregistry.fill(HIST("phi_gen"), phi);
+      QAregistry.fill(HIST("eta_gen"), eta);
+      QAregistry.fill(HIST("pt_gen"), pt);
+    } else {
+      QAregistry.fill(HIST("phi"), phi);
+      QAregistry.fill(HIST("eta"), eta);
+      QAregistry.fill(HIST("pt_dcaXY_dcaZ"), pt, dcaxy, dcaz);
+    }
```

Comment:
For the next iteration consider templating, additionally to the `datatype`) on two objects instead: `maintrack` and `assoctrack` of different `typename`s 
From `maintrack` you always extract `phi`, `eta`, and `pT` and from `assoctrack` you extract the `dcaxy` and `dcaz`
When you call from reconstructed data you call it being both the same
When you call from MC reconstructed you call it with `mcparticle` and with `track`
When you call from MC generated you call it with both the same `mcparticle` it should not complain because you will not use the `assoctrack` 

Something similar can be applied to `setCurrentParticleWeights` then you will avoid introducing the tables row information in the functions calls